provider/block_devices_plug.py

```diff
@@ -0,0 +1,430 @@
+            if self._dev_type != CDROM:
+                if not utils_misc.wait_for(lambda: len(self._imgs) == len(
+                        self._list_all_disks() ^ orig_disks), self._timeout, step=1.5):
+                    raise TestError(
+                        "%s--> Actual: %s disks. Expected: %s disks." %
+                        (action, len(self._all_disks ^ orig_disks), len(self._imgs)))
+                self._plugged_disks = sorted(
+                    [disk.split('/')[-1] for disk in list(self._all_disks ^ orig_disks)])
+            return result
+        return wrapper
+    return decorator
+
+
+class _PlugThread(threading.Thread):
+
+    """
+    Plug Thread that define a plug thread.
+    """
+
+    def __init__(self, vm, action, images, monitor, exit_event, bus=None):
+        threading.Thread.__init__(self)
+        self._action = action
+        self._images = images
+        self._monitor = monitor
+        self._bus = bus
+        self._plug_manager = BlockDevicesPlug(vm)
+        self.exit_event = exit_event
+        self.exc_info = None
+
+    def run(self):
+        try:
+            args = (self._images, self._monitor)
+            if self._bus:
+                args = (self._images, self._monitor, self._bus)
+            method = "_hotplug_devs" if self._action == HOTPLUG else "_unplug_devs"
+            getattr(self._plug_manager, method)(*args)
+        except Exception as e:
+            logging.error(
+                '%s %s failed: %s' % (self._action.capitalize(), self._images, str(e)))
+            self.exc_info = sys.exc_info()
+            self.exit_event.set()
+
+
+class _ThreadManager(object):
+
+    """
+    Thread Manager that provides interfaces about threads for plugging devices.
+    """
+
+    def __init__(self, vm):
+        self._vm = vm
+        self._threads = []
+        self.exit_event = threading.Event()
+
+    def _initial_threads(self, action, imgs, bus=None):
+        max_threads = min(len(imgs), 2 * multiprocessing.cpu_count())
+        for i in xrange(max_threads):
+            mon = self._vm.monitors[i % (len(self._vm.monitors) - 1)]
+            args = (self._vm, action, imgs[i::max_threads], mon, self.exit_event)
+            if bus:
+                args = (self._vm, action, imgs[i::max_threads], mon, self.exit_event, bus)
+            self._threads.append(_PlugThread(*args))
+
+    def _start_threads(self):
+        for thread in self._threads:
+            thread.start()
+
+    def _join_threads(self, timeout):
+        for thread in self._threads:
+            thread.join(timeout)
+
+    def run_threads(self, action, imgs, bus, timeout):
+        self._initial_threads(action, imgs, bus)
+        self._start_threads()
+        self._join_threads(timeout)
+
+    def raise_threads(self):
+        for thread in self._threads:
+            if thread.exc_info:
+                reraise(thread.exc_info[0], thread.exc_info[1], thread.exc_info[2])
```

Comment:
`reraise(*thread.exc_info)`

Reply:
Thanks, will update it, I made such a silly mistake.